### PR TITLE
[GH-1244] Lightweight Testing Dataset(s)

### PR DESCRIPTION
**Pull request by ehigham**
### Purpose
RR: https://broadinstitute.atlassian.net/browse/GH-1244
While TDR can't support snapshotting tables with columns of type `Array<FileId>`, we can't progress directly on supporting the `sarscov2_illumina_full` pipeline. Instead, I'm adding a mock dataset as a generic dataset for ingesting/snapshotting.
I've used this dataset to speed up the `wfl.integration.datarepo-test/test-ingest-pipeline-outputs` test point which now takes significantly less time.
I've added two fake workflows to test traversals and folds of workflow objects. These are:
- `primitive.wdl`: a workflow that demonstrates handling of WDL primitive types
- `compound.wdl`: a workflow that demonstrates handling of WDL compound types

Using these, I've fixed a couple of bugs in my WDL type traversal code. I've also added a prebuild step to get the workflow descriptions from womtool for each WDL under `api/`. This is to remove the need to make extra requests in tests/mock out womtool for unit testing without checking in files.

### CODE REVIEW

File: api/test/resources/workflows/assemble_refbased/assemble_refbased.wdl [hunk 1]
```diff
@@ -0,0 +1 @@
+https://cdn.jsdelivr.net/gh/broadinstitute/viral-pipelines@master/pipes/WDL/workflows/assemble_refbased.wdl
```
Comment (rexwangcc) [inline api/test/resources/workflows/assemble_refbased/assemble_refbased.wdl:1]: should we pin to a commit/tag of the `broadinstitute/viral-pipelines` repo?

File: api/test/resources/workflows/compound.wdl [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    }
+
+    output {
+        Array[String]      outarray       = inarray
+        Map[String, File]  outmap         = inmap
+        File?              outoptional    = inoptional
+        Pair[Int, Float]   outpair        = inpair
+        Struct             outstruct      = instruct
+    }
+}
```
Comment (rexwangcc) [inline api/test/resources/workflows/compound.wdl:30]: (the most well-formatted WDL I've ever worked with 👍 )

File: api/test/resources/workflows/sarscov2_illumina_full/description.edn [hunk 3]
```diff
@@ -1,396 +0,0 @@
-{:valid                   true,
```
Comment (rexwangcc) [inline api/test/resources/workflows/sarscov2_illumina_full/description.edn:1]: nice that we don't have to keep this in

File: api/test/resources/workflows/sarscov2_illumina_full/sarscov2_illumina_full.wdl [hunk 4]
```diff
@@ -0,0 +1 @@
+https://cdn.jsdelivr.net/gh/broadinstitute/viral-pipelines@master/pipes/WDL/workflows/sarscov2_illumina_full.wdl
```
Comment (rexwangcc) [inline api/test/resources/workflows/sarscov2_illumina_full/sarscov2_illumina_full.wdl:1]: same question as the above: should we pin to a commit/tag of the broadinstitute/viral-pipelines repo?

File: api/test/wfl/integration/datarepo_test.clj [hunk 5]
```diff
@@ -31,31 +32,45 @@
...
-        pipeline-outputs (workflows/read-resource "assemble_refbased/outputs")
-        outputs-type     (-> "assemble_refbased/description"
-                             workflows/read-resource
-                             :outputs
-                             workflows/make-object-type)
-        table-name       "assemble_refbased_outputs"
-        workflow-id      (UUID/randomUUID)
-        tdr-profile      (env/getenv "WFL_TDR_DEFAULT_PROFILE")
-        tdr-sa           (env/getenv "WFL_TDR_SA")]
+(def ^:private pi (* 4 (Math/atan 1)))
```
Comment (rexwangcc) [inline api/test/wfl/integration/datarepo_test.clj:35]: Happy Pi Day!

File: api/test/wfl/tools/workflows.clj [hunk 6]
```diff
@@ -1,45 +1,71 @@
...
+       (let [name->type (make-type-environment type)]
+         (reduce-kv #(go %1 (name->type %2) %3) state value))
+       "Optional"
+       (if value (go state (:optionalType type) value) state)
+       "Pair"
+       (let [{:keys [leftType rightType]} (:pairType type)]
+         (-> (go state leftType  (first value))
+             (go       rightType (second value))))
+       (f state (:typeName type) value)))
+   init type object))
```
Comment (rexwangcc) [inline api/test/wfl/tools/workflows.clj:66]: cool

File: api/test/wfl/unit/modules/all_test.clj [hunk 7]
```diff
@@ -11,6 +11,6 @@
   (let [edn          (wfl/get-the-version)
         has-version? #(= (:release %) (-> % :path basename edn))]
     (is (every? has-version? [aou/workflow-wdl
-                              #_sg/workflow-wdl ; Now a commit hash. =tbl
+                              sg/workflow-wdl
```
Comment (rexwangcc) [inline api/test/wfl/unit/modules/all_test.clj:14]: I think it's still a commit hash? https://github.com/broadinstitute/wfl/blob/adac9d5ab0a6872dbabd60697b6400b65b476b67/api/src/wfl/module/sg.clj#L23 @tbl3rd when/ shall we do this here?

**Review by rexwangcc — APPROVED** [review]
Looks good to me in general! I left some questions.

File: api/test/wfl/unit/modules/all_test.clj (hunk 7 above)
Comment (ehigham) [reply api/test/wfl/unit/modules/all_test.clj:14]: Yeah. We do this for wgs as well. I'm not sure why we even include this in "the version".

File: api/test/resources/workflows/assemble_refbased/assemble_refbased.wdl (hunk 1 above)
Comment (ehigham) [reply api/test/resources/workflows/assemble_refbased/assemble_refbased.wdl:1]: Good idea. I'll fix all these :D

File: api/test/resources/workflows/sarscov2_illumina_full/description.edn (hunk 3 above)
Comment (ehigham) [reply api/test/resources/workflows/sarscov2_illumina_full/description.edn:1]: Yep!

File: api/test/wfl/unit/modules/all_test.clj (hunk 7 above)
Comment (rexwangcc) [reply api/test/wfl/unit/modules/all_test.clj:14]: I think for WGS, the reason is cauz I double-submitted what's hot-fixed (which was a commit than a released wdl) in production. I can't remember why we use commit hash for sg, might be that it's still under testing?

Comment (ehigham) [reply api/test/wfl/unit/modules/all_test.clj:14]: Yep. A working WDL hasn't been released yet.

File: api/build/build.clj [hunk 8]
```diff
@@ -75,10 +74,30 @@
           (->> (spit out)))))
   (System/exit 0))
 
+(defn ^:private write-workflow-description [resources wdl]
+  (let [workflow (util/remove-extension (util/basename wdl))
+        file     (io/file resources "workflows" (str workflow ".edn"))]
+    (when-not (.exists file)
+      (println "generating description for" (util/basename wdl))
```
Comment (tbl3rd) [inline api/build/build.clj:81]: Might be handy to report the file this generates.

File: api/test/wfl/integration/datarepo_test.clj (hunk 5 above)
Comment (tbl3rd) [inline api/test/wfl/integration/datarepo_test.clj:35]: Why not just `Math/PI`?

**Pull request merged**